src/libstore/content-address.hh

```diff
@@ -55,23 +89,47 @@ std::optional<ContentAddress> parseContentAddressOpt(std::string_view rawCaOpt);
 
 Hash getContentAddressHash(const ContentAddress & ca);
 
+
 /*
-  We only have one way to hash text with references, so this is single-value
-  type is only useful in std::variant.
-*/
-struct TextHashMethod { };
-struct FixedOutputHashMethod {
-  FileIngestionMethod fileIngestionMethod;
-  HashType hashType;
+ * References set
+ */
+
+struct StoreReferences {
+    StorePathSet others;
+    bool self = false;
+
+    bool empty() const;
+    size_t size() const;
+
+    GENERATE_CMP(StoreReferences, me->self, me->others);
 };
 
-typedef std::variant<
-    TextHashMethod,
-    FixedOutputHashMethod
-  > ContentAddressMethod;
+/*
+ * Full content address
+ *
+ * See the schema for store paths in store-api.cc
+ */
 
-ContentAddressMethod parseContentAddressMethod(std::string_view rawCaMethod);
+// This matches the additional info that we need for makeTextPath
+struct TextInfo : TextHash {
```

Comment:
I think using inheritance here isn't really nice. It's easy-enough to confuse one for another without adding some implicit conversion between them. Can't we have a proper subfield instead?

```suggestion
struct TextInfo {
    TextHash hash;
```

(same goes for [FixedOutputInfo](https://github.com/NixOS/nix/pull/3746/files#diff-42e44a338fd5db8e9cd1f43e4522641e0fe8300bff8d58a8c7edd6ac82f827e8R121) ofc)

Reply:
Also, that would allow properly using aggregate initialization, which would be a huge win for readability in the rest of the code

Reply:
Yeah when I first did this, I could use flat designated initializers in GCC (in a non-standard way which I didn't realize) which was nice. Now absolutely the inheritance is not pulling its weight.

Reply:
Done. I wonder also if it might make sense to just inline the type to avoid `hash.hash` and make the designated initializers nicer.

Kind of an evil thought (flattening and duplicating data definitions for sake of nice syntax), but very tempting :)